File: models/Gpt-3.5-turbo/cpp/code/problems/graph_algorithms/Knight's_Tour.cpp

#include <iostream>
#include <vector>

using namespace std;

// Global variables to represent the chessboard and the current knight position
vector<vector<int>> board(8, vector<int>(8, 0));
int dx[] = {2, 2, -2, -2, 1, 1, -1, -1};
int dy[] = {1, -1, 1, -1, 2, -2, 2, -2};

// Function to check if a move is valid
bool isValidMove(int x, int y) {
    return (x >= 0 && x < 8 && y >= 0 && y < 8 && board[x][y] == 0);
}

// Recursive function to find the knight's tour
bool findKnightTour(int x, int y, int count) {
    // Mark the current square as visited
    board[x][y] = count;

    // Base case: if all squares are visited, return true
    if (count == 64) {
        return true;
    }

    // Try all possible moves from the current position
    for (int i = 0; i < 8; i++) {
        int nextX = x + dx[i];
        int nextY = y + dy[i];

        // Check if the next move is valid
        if (isValidMove(nextX, nextY)) {
            // Recursive call to explore the next move
            if (findKnightTour(nextX, nextY, count + 1)) {
                return true;
            }
        }
    }

    // If no valid move is found, backtrack and reset the current square
    board[x][y] = 0;
    return false;
}

int main() {
    int x, y;
    cin >> x >> y;

    // Start the knight's tour from the given position
    findKnightTour(x - 1, y - 1, 1);

    // Print the knight's tour
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            cout << board[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/graph_algorithms/Knight's_Tour.py

def is_valid_move(x, y):
    return 1 <= x <= 8 and 1 <= y <= 8

def dfs(x, y, count, visited, moves):
    # Mark the current square as visited
    visited[x][y] = True
    
    # Add the current square to the moves array
    moves.append((x, y))
    
    # If all squares have been visited, return True
    if count == 64:
        return True
    
    # Possible knight moves
    dx = [2, 1, -1, -2, -2, -1, 1, 2]
    dy = [1, 2, 2, 1, -1, -2, -2, -1]
    
    # Try all possible moves from the current square
    for i in range(8):
        new_x = x + dx[i]
        new_y = y + dy[i]
        
        # Check if the move is valid and the square has not been visited yet
        if is_valid_move(new_x, new_y) and not visited[new_x][new_y]:
            # Recursive call to explore the next square
            if dfs(new_x, new_y, count+1, visited, moves):
                return True
    
    # Backtrack: remove the last move from the moves array and mark the current square as unvisited
    moves.pop()
    visited[x][y] = False
    
    return False

# Read the input
x, y = map(int, input().split())

# Create the visited array
visited = [[False] * 9 for _ in range(9)]

# Create the moves array
moves = []

# Call the DFS function to find the sequence of moves
dfs(x, y, 1, visited, moves)

# Print the moves array
for move in moves:
    print(move[0], move[1], end=' ')
